index.js

```diff
@@ -561,6 +570,43 @@ function buildArray (schema, code, name, externalSchema, fullSchema) {
   return code
 }
 
+function buildArrayTypeCondition (type, accessor) {
+  var condition
+  switch (type) {
+    case 'null':
+      condition = `obj${accessor} === null`
+      break
+    case 'string':
+      condition = `typeof obj${accessor} === 'string'`
+      break
+    case 'integer':
+      condition = `Number.isInteger(obj${accessor})`
+      break
+    case 'number':
+      condition = `Number.isFinite(obj${accessor})`
+      break
+    case 'boolean':
+      condition = `typeof obj${accessor} === 'boolean'`
+      break
+    case 'object':
+      condition = `obj${accessor} && typeof obj${accessor} === 'object' && obj${accessor}.constructor === Object`
+      break
+    case 'array':
+      condition = `Array.isArray(obj${accessor})`
+      break
+    default:
+      if (Array.isArray(type)) {
+        var conditions = type.map((subType) => {
+          return buildArrayTypeCondition(subType, accessor)
+        })
+        condition = `(${conditions.join(' || ')})`
```

Comment:
There is potential for an optimization here. In the current implementation the generated code checks if the item matches any of the types. The `nested()` basically performs the same check again using ajv. We could split this condition here into separate ones and use the knowledge of the item's type for the `nested()` function. However, this requires some more refactoring.

Reply:
Feel free to experiment! We do like optimizations! :D

Reply:
I'm landing this now, feel free to send a followup PR with the optimizations!